### PR TITLE
Ta feedback feature

### PR DESCRIPTION
**Pull request by warlordwinter**
# Here is all of the code related to TA-Feedback

### CODE REVIEW

File: config.json [hunk 1]
```diff
@@ -3,6 +3,9 @@
     "command_channels": [
       1192895746672951340
     ],
+    "ta_review_channel": [
+      1284224818698260490
+    ],
```
Comment (byubean) [inline config.json:8]: We need the TA review channel to be server specific.

File: discord_bot.py [hunk 2]
```diff
@@ -126,17 +132,39 @@ def __init__(self, root_save_folder: Path, config):
 
         state_folder = root_save_folder / 'history'
         metrics_folder = root_save_folder / 'metrics'
-
         # MetricsHandler initialization
         self.metrics_handler = MetricsHandler(metrics_folder)
 
+        if self._ta_channel_config is None:
+            # TODO - Exception
+            print("ta_review_channel is not set in the configuration.")
```
Comment (byubean) [inline discord_bot.py:140]: I think this all can go.

File: discord_bot.py [hunk 3]
```diff
@@ -149,9 +177,8 @@ async def on_ready(self):
         logging.info(self.user.id)
         logging.info('Starting workflow manager')
         self._workflow_manager = await self._workflow_manager.__aenter__()
-        await asyncio.sleep(0.1)
-        logging.info('Workflow manager ready')
 
+        #Create a ta
```
Comment (byubean) [inline discord_bot.py:181]: delete this comment

File: prompts/standard-rubber-duck.txt [hunk 4]
```diff
@@ -18,4 +18,6 @@ As an AI CS instructor:
 - if a student asks you to write code to solve a problem, do not; instead, invite them to try and encourage them step-by-step without telling them what the next step is
 - if a student provides ideas that don't match the instructions they may have shared, ask questions that help them achieve greater clarity
 - sometimes students will resist coming up with their own ideas and want you to do the work for them; however, after a few rounds of gentle encouragement, a student will start trying. This is the goal.
-- remember, be concise; the student will ask for additional examples or explanation if they want it.
\ No newline at end of file
+- remember, be concise; the student will ask for additional examples or explanation if they want it.
+- if giving them any code, generalize it so that it doesn't apply to any specifics they've requested and keep it simple.
+- You can also add links to official python/cpp/etc documentation of how to code whatever tasks they're asking for. Ask them if they'd like a link to documentation after providing code snippets.
```
Comment (byubean) [inline prompts/standard-rubber-duck.txt:23]: Did you guys add this to the prompt?

File: pyproject.toml [hunk 5]
```diff
@@ -9,6 +9,7 @@ python = "^3.11"
 openai = "^1.11.1"
 discord = "^2.1.0"
 urllib3 = "<2.0"
+seaborn = "0.13.2"
```
Comment (byubean) [inline pyproject.toml:12]: I'm guessing this is for Bryce's branch?

File: rubber_duck.py [hunk 6]
```diff
@@ -106,15 +106,11 @@ def __init__(self,
...
-        self.feedback_workflow = FeedbackWorkflow(
-            metrics_handler.record_feedback,
-            workflow_manager.send_event,
-            message_handler.send_message
-        )
+        self.start_feedback_workflow = step(start_feedback_workflow)
+
 
-    async def __call__(self, thread_id: int, engine: str, prompt: str, initial_message: Message, timeout=600):
+    async def __call__(self, thread_id: int, engine: str, prompt: str, initial_message: Message, timeout=60):#we changed this to 60 seconds for testing purposes
```
Comment (byubean) [inline rubber_duck.py:113]: Revert this change and remove the comment

File: rubber_duck.py [hunk 7]
```diff
@@ -131,7 +127,7 @@ def generate_error_message(self, guild_id, thread_id, ex):
     #
     # Begin Conversation
     #
-    async def have_conversation(self, thread_id: int, engine: str, prompt: str, initial_message: Message, timeout=600):
+    async def have_conversation(self, thread_id: int, engine: str, prompt: str, initial_message: Message, timeout=20): #we changed this to 60 seconds.
```
Comment (byubean) [inline rubber_duck.py:130]: same here

File: rubber_duck.py [hunk 8]
```diff
@@ -164,6 +160,10 @@ async def have_conversation(self, thread_id: int, engine: str, prompt: str, init
                         )
                         continue
 
+                    # Here, we could check the similarity between the current and previous messages, and prompt them to
+                    # start a new thread if there is a low similarity
+
+
```
Comment (byubean) [inline rubber_duck.py:166]: Let's remove this for now

File: rubber_duck.py [hunk 9]
```diff
@@ -258,3 +258,4 @@ async def _get_completion_with_retry(self, thread_id, engine, message_history):
                     f"Retrying due to {ex}, attempt {retries}/{max_retries}. Waiting {delay} seconds.")
                 await asyncio.sleep(delay)
                 delay *= backoff
+#Add a function that removes the state folder every run.
```
Comment (byubean) [inline rubber_duck.py:261]: Remove comment

File: prompts/standard-rubber-duck.txt (hunk 4 above)
Comment (warlordwinter) [reply prompts/standard-rubber-duck.txt:23]: Yes we did add that. Do you want that taken out?

File: pyproject.toml (hunk 5 above)
Comment (warlordwinter) [reply pyproject.toml:12]: Yes!

File: config.json (hunk 1 above)
Comment (warlordwinter) [reply config.json:8]: Okay I'm going to delete that part and leave it empty for the user to insert in their server

File: discord_bot.py (hunk 2 above)
Comment (warlordwinter) [reply discord_bot.py:140]: Done

File: pyproject.toml (hunk 5 above)
Comment (warlordwinter) [reply pyproject.toml:12]: I will remove it for now and when Bryce merges we can add that in.

File: rubber_duck.py (hunk 6 above)
Comment (warlordwinter) [reply rubber_duck.py:113]: Done

File: rubber_duck.py (hunk 9 above)
Comment (warlordwinter) [reply rubber_duck.py:261]: done

**Pull request merged**